backend/groth16/bls12-377/mpcsetup/phase1.go

```diff
@@ -6,187 +6,254 @@
+		// let α₁ = α₀.α', τ₁ = τ₀.τ'
+		// then α₁τ₁ⁱ = (α₀τ₀ⁱ)α'τ'ⁱ
+		alphaUpdates[i].Mul(&tauUpdates[i], alphaUpdate)
 	}
-	if !sameRatio(contribution.PublicKeys.Alpha.SG, contribution.PublicKeys.Alpha.SXG, contribution.PublicKeys.Alpha.XR, alphaR) {
-		return errors.New("couldn't verify public key of α")
+	scaleG1InPlace(c.G1.AlphaTau, alphaUpdates)
+
+	betaUpdates := make([]fr.Element, len(c.G1.BetaTau))
+	betaUpdates[0].Set(betaUpdate)
+	for i := range betaUpdates {
+		betaUpdates[i].Mul(&tauUpdates[i], betaUpdate)
 	}
-	if !sameRatio(contribution.PublicKeys.Beta.SG, contribution.PublicKeys.Beta.SXG, contribution.PublicKeys.Beta.XR, betaR) {
-		return errors.New("couldn't verify public key of β")
+	scaleG1InPlace(c.G1.BetaTau, betaUpdates)
+
+	var betaUpdateI big.Int
+	betaUpdate.BigInt(&betaUpdateI)
+	c.G2.Beta.ScalarMultiplication(&c.G2.Beta, &betaUpdateI)
+}
+
+// Seal performs the final contribution and outputs the final parameters.
+// No randomization is performed at this step.
+// A verifier should simply re-run this and check
+// that it produces the same values.
+// The inner workings of the random beacon are out of scope.
+// WARNING: Seal modifies p, just as Contribute does.
+// The result will be an INVALID Phase1 object, since no proof of correctness is produced.
+func (p *Phase1) Seal(beaconChallenge []byte) SrsCommons {
+	newContribs := mpcsetup.BeaconContributions(p.hash(), []byte("Groth16 MPC Setup - Phase 1"), beaconChallenge, 3)
+	p.parameters.update(&newContribs[0], &newContribs[1], &newContribs[2])
+	return p.parameters
+}
+
+// VerifyPhase1 and return the SRS parameters usable for any circuit of domain size N
+// beaconChallenge is the output of the random beacon
+// and c are the output from the contributors
+// WARNING: the last contribution object will be modified
+func VerifyPhase1(N uint64, beaconChallenge []byte, c ...*Phase1) (SrsCommons, error) {
+	prev := NewPhase1(N)
+	wp := gcUtils.NewWorkerPool()
+	defer wp.Stop()
+	for i := range c {
+		if err := prev.Verify(c[i], WithWorkerPool(wp)); err != nil {
+			return SrsCommons{}, err
+		}
+		prev = c[i]
 	}
+	return prev.Seal(beaconChallenge), nil
+}
+
+// Verify assumes previous is correct
+func (p *Phase1) Verify(next *Phase1, options ...verificationOption) error {
 
-	// Check for valid updates using previous parameters
-	if !sameRatio(contribution.Parameters.G1.Tau[1], current.Parameters.G1.Tau[1], tauR, contribution.PublicKeys.Tau.XR) {
-		return errors.New("couldn't verify that [τ]₁ is based on previous contribution")
+	challenge := p.hash()
+	if len(next.Challenge) != 0 && !bytes.Equal(next.Challenge, challenge) {
+		return errors.New("the challenge does not match the previous contribution's hash")
 	}
-	if !sameRatio(contribution.Parameters.G1.AlphaTau[0], current.Parameters.G1.AlphaTau[0], alphaR, contribution.PublicKeys.Alpha.XR) {
-		return errors.New("couldn't verify that [α]₁ is based on previous contribution")
+	next.Challenge = challenge
```

Comment:
My assumption was that verification wouldn't change the inputs, but it does as we write `next.Challenge`. This imo may have practical side effects that the contributions are not static (a la when hashing serialized proofs). Instead, I would provide the challenge as argument to `Verify` and then `VerifyPhase1` automatically carries the continuation over to next round (with computing it using `newChallenge := Phase1.Hash(challenge)`)

Reply:
It's only changed if it is empty.